packages/vitest/src/node/reporters/renderers/dotRenderer.ts

```diff
@@ -27,54 +27,68 @@ function getIcon(task: Task) {
   }
 }
 
-function render(tasks: Task[]): string {
+function render(tasks: Task[], width: number): string {
   const all = getTests(tasks)
-  const output: string[] = []
+  let currentIcon = pending
+  let currentTasks = 0
+  let previousLineWidth = 0
+  let output = ''
 
   // The log-update uses various ANSI helper utilities, e.g. ansi-warp, ansi-slice,
   // when printing. Passing it hundreds of single characters containing ANSI codes reduces
   // performances. We can optimize it by reducing amount of ANSI codes, e.g. by coloring
   // multiple tasks at once instead of each task separately.
-  let currentIcon = pending
-  let currentTasks = 0
-
-  const addOutput = () => output.push(currentIcon.color(currentIcon.char.repeat(currentTasks)))
+  const addOutput = () => {
+    const { char, color } = currentIcon
+    const availableWidth = width - previousLineWidth
+    if (availableWidth > currentTasks) {
+      output += color(char.repeat(currentTasks))
+      previousLineWidth += currentTasks
+    }
+    else {
+      // We need to split the line otherwise it will mess up log-update's height calculation
+      // and spam the scrollback buffer with dots.
+      let buf = `${char.repeat(availableWidth)}\n`
+      let remaining = currentTasks - availableWidth
+      buf += `${char.repeat(width)}\n`.repeat(Math.floor(remaining / width))
+      remaining %= width
+
+      if (remaining > 0) {
+        buf += char.repeat(remaining)
+        previousLineWidth = remaining
+      }
+      else {
+        previousLineWidth = 0
+      }
```

Comment:
While this works fine I found it a bit difficult to read. I had to do some refactoring to it in order to understand what exactly is happening. I came up with following refactoring in the end. Would it be more readable?

```suggestion
      let buf = `${char.repeat(availableWidth)}\n`
      const remaining = currentTasks - availableWidth

      // Add full rows first
      const fullRows = Math.floor(remaining / width)
      if (fullRows > 0)
        buf += `${char.repeat(width)}\n`.repeat(fullRows)

      // Add remaining dots which don't make a full row
      const partialRow = remaining % width
      if (partialRow > 0) {
        buf += `${char.repeat(partialRow)}`
        previousLineWidth = partialRow
      }
      else {
        previousLineWidth = 0
      }
```

Reply:
Sounds good. I've pushed the suggested changes with a few tweaks.